py/htm/encoders/eye.py

```diff
@@ -0,0 +1,596 @@
+                 range of possible channel values.
+
+        Argument wrap ... default is False.
+                 This supports modular input spaces and ranges which wrap
+                 around. It does this by rotating the inputs by a constant
+                 random amount which hides where the discontinuity in ranges is.
+                 No ranges actually wrap around the input space.
+        """
+        self.input_shape  = tuple(input_shape)
+        self.num_samples  = int(round(num_samples))
+        self.sparsity     = sparsity
+        self.output_shape = self.input_shape + (self.num_samples,)
+        self.dtype        = dtype
+        self.drange       = drange
+        self.len_drange   = max(drange) - min(drange)
+        self.wrap         = bool(wrap)
+        if self.wrap:
+            # Each bit responds to a range of input values, length of range is 2*Radius.
+            radius        = self.len_drange * self.sparsity / 2
+            self.offsets  = np.random.uniform(0, self.len_drange, self.input_shape)
+            self.offsets  = np.array(self.offsets, dtype=self.dtype)
+            # If wrapping is enabled then don't generate ranges which will be
+            # truncated near the edges.
+            centers = np.random.uniform(min(self.drange) + radius,
+                                        max(self.drange) - radius,
+                                        size=self.output_shape)
+        else:
+            # Buckets near the edges of the datarange are OK.  They will not
+            # respond to a full range of input values but are needed to
+            # represent the bits at the edges of the data range.
+            #
+            # Expand the data range and create bits which will encode for the
+            # edges of the datarange to ensure a resonable sparsity at the
+            # extremes of the data ragne.  Increase the size of all of the
+            # buckets to accomidate for the extra area being represented.
+            M = .95 # Maximum fraction of bucket-size outside of the true data range to allocate buckets.
+            len_pad_drange = self.len_drange / (1 - M * self.sparsity)
+            extra_space    = (len_pad_drange - self.len_drange) / 2
+            pad_drange     = (min(self.drange) - extra_space, max(self.drange) + extra_space)
+            radius         = len_pad_drange * self.sparsity / 2
+            centers = np.random.uniform(min(pad_drange),
+                                        max(pad_drange),
+                                        size=self.output_shape)
+        # Make the lower and upper bounds of the ranges.
+        low  = np.clip(centers - radius, min(self.drange), max(self.drange))
+        high = np.clip(centers + radius, min(self.drange), max(self.drange))
+        self.low  = np.array(low, dtype=self.dtype)
+        self.high = np.array(high, dtype=self.dtype)
+
+    def encode(self, img):
```

Comment:
done for output, had some troubles with input as SDR, so I'm leaving that for later. Anyway, Channel is tightly coupled with Retina/image data - expects local similarity, log-pol encoding, ... 
so I think it's fine to expect an image as input here (?) 

Actually, I was thinking it'd be good if the log-pol transform was a part of this encoder (so Eye can be purely bioinspired.Retina) ...?

Reply:
> Actually, I was thinking it'd be good if the log-pol transform was a part of this encoder (so Eye can be purely bioinspired.Retina) ...?

I disagree. The retina does should do the log-polar transform, and the channel encoder should only do the transform from `pixels -> SDR`.  The Channel encoder could be used with other types of sensory data.

Reply:
> The Channel encoder could be used with other types of sensory data.

are there any requirements on the data (similarity, encoding) for Channel to be used? Ie can I use it as a replacement for RDSE for encoding floats?

Reply:
This encoder needs to know the data range, so it can not replace the RDSE.  It could in theory replace the scalar encoder but I would not recommend it, since the scalar encoder works great and has nice deterministic properties.

Reply:
ok, I'll make the split to standalone file + test (maybe as a part of this PR?)